Rewrite _geometric docstrings for SK2 xy-ij change

matthew-brett opened the pull request:
Rewrite docstrings in attempt to make them work for both
`skimage.transform.warp`, and the `skimage2` `warp`.

## Description

Attempt to generalize docstrings so they make sense either as `ij` or as `xy`
transformations, ready for `skimage2`.

<!--
- Reference relevant issues or related pull requests with their URL / #<number>.
- Use `pre-commit` to check and format code.
-->

## Checklist

<!-- Before pull requests can be merged, they should provide: -->

- A descriptive but concise pull request title
- [Docstrings for all functions](https://github.com/numpy/numpydoc/blob/main/doc/example.py)
- [Unit tests](https://scikit-image.org/docs/dev/development/contribute.html#testing)
- A gallery example in `./doc/examples` for new features
- [Contribution guide](https://scikit-image.org/docs/dev/development/contribute.html) is followed

## Release note

For maintainers and optionally contributors, please refer to the [instructions](https://scikit-image.org/docs/stable/development/contribute.html#documenting-changes) on how to document this PR for the release notes.

```release-note
...
```

matthew-brett committed:
Rewrite _geometric docstrings for SK2 xy-ij change

Rewrite docstrings in attempt to make them work for both
`skimage.transform.warp`, and the `skimage2` `warp`.

stefanv commented on src/skimage/transform/_geometric.py:
```diff
@@ -934,26 +934,50 @@ def estimate(self, src, dst):
     Apply a projective transformation (homography) on coordinates.
 
-    For each homogeneous coordinate :math:`\mathbf{x} = [x, y, 1]^T`, its
+    In what follows, ``x`` refers to coordinates from the first coordinate
```
This is pretty confusing, given that we now have both `xy` and `ij` conventions. But, I don't see a better way.

matthew-brett replied:
No, here I trying to point out that this is a general description of the transform, where $x$ in the description refers to the first axis (whatever that is) and $y$ in the description refers to the second axis (whatever that is).  In Skimage2, we can rewrite to use $i$ and $j$ if we want, but we have to leave it this way for Skimage1 ...

stefanv commented on src/skimage/transform/_geometric.py:
```diff
@@ -1341,14 +1376,15 @@ class AffineTransform(ProjectiveTransform):
-    2. Shear
-    3. Rotate
-    4. Translate
+    1. Scale;
```
the semi-colons look weird to me :shrug:

stefanv reviewed (commented):
I think this is a pretty good middle ground. I recommended some edits to the new notes.

matthew-brett committed:
Remove punctuation at end of single-word list items

matthew-brett committed:
Use and extend Stéfan's rewrite of rotation note

matthew-brett committed:
Apply suggestions from code review

Co-authored-by: Lars Grüter <lagru+github@mailbox.org>

matthew-brett commented on src/skimage/transform/_geometric.py:
```diff
@@ -934,26 +934,50 @@ def estimate(self, src, dst):
+    Whether this will appear as a clockwise or counterclockwise rotation will
+    depend on how you apply the transform to the image; see the Notes section.
+
+    To translate x by 10 and y by 20::
```
Only to say that, I suppose one could use $x$ and $y$ instead, to signify these as mathematical ideas, but maybe that wouldn't be any clearer.

stefanv reviewed (approved):
I left some minor points, but overall looks good to me; thanks @matthew-brett. Feel free to merge after addressing outstanding points to your satisfaction.

matthew-brett committed:
Apply suggestions from code review

Co-authored-by: Stefan van der Walt <stefan@mentat.za.net>

lagru reviewed (changes requested):
I don't feel good in holding this up further, but I feel that it's too early to include the notice about the convention change between v1 and v2.

Other comments aren't blocking but it would be good to clarify those. Maybe I'm being just a bit dense.

matthew-brett committed:
Merge remote-tracking branch 'origin/main' into geom-docs

* origin/main: (254 commits)
  Refactor, document migration utilities (#8184)
  ci: improve handling of pr labels (#8169)
  Doctest migration in PRs & on all platforms (#8141)
  Update SKIP 4 (import hierarchy, backporting, testing) (#8097)
  One page per function / class in HTML docs for skimage2 (#8154)
  Pin minimum pytest to 9 (#8174)
  docs: make source install instructions more discoverable and add build requirements to top level requirements file (#8165)
  Drop `--maxfail=5` from local pytest config (#8170)
  Use standard C99 static inline instead of `__inline` (#8159)
  Fix structural_similarity win_size / gaussian_weights conflict (#8162)
  Unwrap seed refactor (#8142)
  ci: Remove llvm build and add ccache support on windows (#8164)
  Document warn_external in contributing guide (#8163)
  CI: Fix emscripten dependencies and clean up wheel test commands (#8147)
  Test only changed subpackages (#8103)
  Fix nightlies: update deprecated scipy call, fix underdetermined test
  Fix minkowski distanced invocation for SciPy >= 1.18 (#8157)
  Port `structural_similarly` to ski2 (#8104)
  make spin build requirement a lower bound (#8158)
  Drop Python 3.11 Support (#8153)
  ...

matthew-brett commented on src/skimage/transform/_geometric.py:
```diff
@@ -934,26 +934,50 @@ def estimate(self, src, dst):
+_WARP_NOTES = '''Notes
+    -----
+
+    Our axis convention will change from Scikit-image 1 to 2, and this changes the
```
Then - can I ask your advice - when and how should this text be shown?

Options I can think of:

* As currently, in the notes, as warning for an upcoming change.  I am leaning towards this, as a) our next release will have skimage2 (won't it?), and it seems to me it does no harm here.
* As a migration docstring.   Then the warning only appears whenever we enable it, and presumably, it appears as a deprecation warning.  But - it isn't quite that - it's general information about changes in behavior elsewhere - and specifically in `warp`.   We'd also probably have to do something to prevent the text appearing in the migration guide for each function.

What do you think?

stefanv replied:
I think this text, or some version of it, should appear in 1.0, so users know what to expect. In all general locations, like `warp`, where it applies, is good.

lagru replied:
I'm okay with including it then.

matthew-brett commented on src/skimage/transform/_geometric.py:
```diff
@@ -934,26 +934,50 @@ def estimate(self, src, dst):
+    Note, therefore, that if you apply a transform using ``warp`` and other
+    functions applying transforms, from ``skimage2``, you will get an opposite
+    direction rotation from the same function in the ``skimage.transform``
+    module.'''
```
@stefanv - what do you think about putting a note in the migration guide for this, and leave until a bit later, when I'm porting `warp`?

stefanv replied:
That works!

matthew-brett commented on src/skimage/transform/_geometric.py:
```diff
@@ -934,26 +934,50 @@ def estimate(self, src, dst):
+    In what follows, ``x`` refers to coordinates from the first coordinate
+    axis, and ``y`` refers to coordinates from the second coordinate axis.
+
+    For each homogeneous coordinate :math:`\mathbf{{x}} = [x, y, 1]^T`, its
```
I've rewritten to be (I hope) clearer.

matthew-brett commented on src/skimage/transform/_geometric.py:
```diff
@@ -934,26 +934,50 @@ def estimate(self, src, dst):
 
       [[a0 a1 a2]
        [b0 b1 b2]
        [c0 c1 1 ]].
```
I applied this in a recent commit.

matthew-brett commented on src/skimage/transform/_geometric.py:
```diff
@@ -934,26 +934,50 @@ def estimate(self, src, dst):
        [c0 c1 1 ]].
 
-    E.g., to rotate by theta degrees clockwise, the matrix should be::
+    E.g., to rotate by theta degrees, the matrix should be::
```
Applied in recent commit.

matthew-brett commented on src/skimage/transform/_geometric.py:
```diff
@@ -1848,11 +1887,12 @@ class EuclideanTransform(ProjectiveTransform):
-        pass that as `matrix` above.
+        Rotation angle, in radians. See Notes for discussion of clockwise vs
+        anti-clockwise.  If given as a vector, it is interpreted as Euler
+        rotation angles [1]_. Only 2D (single rotation) and 3D (Euler
```
OK to pretend that 3D doesn't exist for now?

stefanv replied:
Yes, I think that's fine; covers most cases, and 3D, if we are consistent, will be easy to track once 2D is understood.

matthew-brett committed:
Rewrite geometric docstring

matthew-brett reviewed (commented):
Trying to work out the best way to display the warning.   Some edits to explanation.

matthew-brett committed:
Apply suggestions from code review

Co-authored-by: greptile-apps[bot] <165735046+greptile-apps[bot]@users.noreply.github.com>

stefanv commented on src/skimage/transform/_geometric.py:
```diff
@@ -934,26 +934,63 @@ def estimate(self, src, dst):
+    Whether this will appear as a counter-clockwise or clockwise rotation in
+    terms of an image array, will depend on your input coordinates; if the
+    first column in your coordinate array has coordinates in terms of the
+    first image (array) axis, then the ``x`` coordinates above correspond to
```
this can accidentally be read as "'first image'-axis". Not sure about the best fix, but duplication wouldn't be the worst: "first image axis (first array axis)"?

stefanv replied:
The second mention below can probably stay as is, since the user then knows what we're talking about.

matthew-brett replied:
Done.

stefanv commented on src/skimage/transform/_geometric.py:
```diff
@@ -934,26 +934,63 @@ def estimate(self, src, dst):
+    first column in your coordinate array has coordinates in terms of the
+    first image (array) axis, then the ``x`` coordinates above correspond to
+    the first (row, ``i``) axis in the image array. In that case the rotation
+    applied to the coordinates is *counter-clockwise*. If the first column has
```
> If the first column has
>    coordinates from the second image array axis

Hard to parse, can you think of a way to make this clearer?

matthew-brett replied:
I have tried to do that.

stefanv commented on src/skimage/transform/_geometric.py:
```diff
@@ -934,26 +934,63 @@ def estimate(self, src, dst):
+    *clockwise*.
 
-    or, to translate x by 10 and y by 20::
+    To translate ``x`` (first coordinate axis) by 10 and ``y`` by 20::
```
Now we're talking about x and y again? :think:

matthew-brett replied:
Yes, this was essentially a trick to try and make the docstrings make sense for `skimage` and `skimage2`.  I'm using `x` to generically refer to the coordinates in the first column of the coordinate array.   I'm hoping that, for users of `skimage`, they will think - oh you mean `x` as in the columns, and the users of `skimage2` will think - x doesn't mean anything in terms of coordinates, so it's just a generic label for a variable.

stefanv commented on src/skimage/transform/_geometric.py:
```diff
@@ -1081,6 +1120,9 @@ def from_estimate(cls, src, dst, weights=None):
 
         Number of source and destination coordinates must match.
 
+        In what follows, ``x`` refers to coordinates from the first coordinate
```
Aren't these what we decided to call `i` and `j`? Or is this to try and stick with the literature x and y?

matthew-brett replied:
Yes, as above.

matthew-brett committed:
Attempt rewrite of docstring for clarity.

matthew-brett reviewed (commented):
Comments on comments.

lagru commented on src/skimage/transform/_geometric.py:
```diff
@@ -934,26 +934,63 @@ def estimate(self, src, dst):
         return self._estimate(src, dst) is None
 
 
+_WARP_NOTES = '''Notes
+    -----
```
Maybe this shouldn't be a _Notes_ section but rather an admonition? Maybe even one that is collapsed by default (for now)?

stefanv commented on src/skimage/transform/_geometric.py:
```diff
@@ -934,26 +934,63 @@ def estimate(self, src, dst):
+_WARP_NOTES = '''Notes
+    -----
+
+    Our axis convention will change from Scikit-image 1 to 2, and this changes the
```
```suggestion
    Our axis convention will change from scikit-image 1 to 2, and this changes the
```

matthew-brett commented:
We have changed the base branch for this PR from the `main` branch to the `pre-megamove-integration`.

## Summary - what this means for your PR

You can continue to work on your PR, now based on `pre-megamove-integration`.  You should update your branch to the current state of `pre-megamove-integration` (this is what the Github "Update branch" button does).   Please **do not merge the scikit-image `main` branch into your PR branch**.

Once we've merged your PR into the `pre-megamove-integration` branch, we'll update your changes for Megamove, and merge into the `main` branch.

Read on for the rationale, and more detailed instructions.

<details>

## Rationale

We've changed the base branch to make it easier and smoother to continue to work on your PR, while we work through a rather large change in the directory structure of the repository - a change that we have called *Megamove*.

See [the Megamove explanation issue](https://github.com/scikit-image/scikit-image/issues/8232) for more detail.

Changing the base branch means you do not see merge conflicts related to the *Megamove* change, and can continue to work with the repository in its original state, without worrying about porting those changes across to the new post-Megamove structure.

As you'll see in the linked issue, you can continue to work on your PR as before, and you can continue to use the Github GUI to "Update branch" - this updates from the new `pre-megamove-integration` branch.   However, there is one caveat: please, do not try and merge the `scikit-image` `main` branch into this one, but instead, keep track of the `pre-megamove-integration` branch.  The issue explains more.

To emphasize again, **do not try and merge the scikit-image `main` branch**.  You will get Megamove related merge conflicts that are rather tricky to resolve correctly.

When we've approved your PR, we'll merge into the `pre-megamove-integration` branch, and then process it so we can merge it into the `main` branch.   We'll let you know when we've merged your PR changes into `main`.

</details>

Any questions, please reply here.
